Sources/NIO/Codec.swift

```diff
@@ -790,3 +790,270 @@ extension MessageToByteHandler {
+///                 _ = context.writeAndFlush(self.wrapOutboundOut(HTTPServerResponsePart.end(nil)))
+///             }
+///         }
+///     }
+///
+///     let group = MultiThreadedEventLoopGroup(numberOfThreads: 1)
+///     let bootstrap = ServerBootstrap(group: group).childChannelInitializer({channel in
+///         channel.pipeline.configureHTTPServerPipeline(withPipeliningAssistance: true, withErrorHandling: true).flatMap { _ in
+///             channel.pipeline.addHandlers([RawBodyMessageHandler(), DecodedBodyHTTPHandler()])
+///         }
+///     })
+///     let channelFuture = bootstrap.bind(host: "127.0.0.1", port: 0)
+///
+public class NIOSingleStepByteToMessageProcessor<Decoder: NIOSingleStepByteToMessageDecoder> {
+    typealias Decoder = Decoder
+
+    private enum DecodeMode {
+        /// This is a usual decode, ie. not the last chunk
+        case normal
+        /// Last chunk
+        case last
+    }
+
+    private var decoder: Decoder
+    private let maximumBufferSize: Int?
+    internal private(set) var buffer: ByteBuffer?
+
+    /// Initialize a `NIOSingleStepByteToMessageProcessor`.
+    ///
+    /// - parameters:
+    ///     - decoder: The `NIOSingleStepByteToMessageDecoder` to decode the bytes into message.
+    ///     - maximumBufferSize: The maximum number of bytes to aggregate in-memory.
+    public init(_ decoder: Decoder, maximumBufferSize: Int? = nil) {
+        self.decoder = decoder
+        self.maximumBufferSize = maximumBufferSize
+    }
+
+    private func append(_ buffer: ByteBuffer) {
+        if self.buffer == nil {
+            self.buffer = buffer
+        } else {
+            var buffer = buffer
+            self.buffer!.writeBuffer(&buffer)
+        }
+    }
+
+    private func decodeLoop(decodeMode: DecodeMode, seenEOF: Bool = false, _ messageReceiver: (Decoder.InboundOut) -> Void) throws {
+        // we want to call decodeLast once with an empty buffer if we have nothing
+        if decodeMode == .last && (self.buffer == nil || self.buffer!.readableBytes == 0) {
+            var emptyBuffer = self.buffer == nil ? ByteBufferAllocator().buffer(capacity: 0) : self.buffer!
+            let message = try self.decoder.decodeLast(buffer: &emptyBuffer, seenEOF: seenEOF)
+            if message != nil {
+                messageReceiver(message!)
+            }
+            return
+        }
+
+        // buffer can only be nil if we're called from finishProcessing which is handled above
+        assert(self.buffer != nil)
+
+        var decodingState: DecodingState = .continue
+        while self.buffer!.readableBytes > 0 && decodingState == .continue {
+            let message: Decoder.InboundOut?
+            if decodeMode == .normal {
+                message = try self.decoder.decode(buffer: &self.buffer!)
+                if message == nil, let maximumBufferSize = self.maximumBufferSize, self.buffer!.readableBytes > maximumBufferSize {
```

Comment:
That's not unreasonable, but I was attempting to copy the behavior of `ByteToMessageHandler`. It allows storing more than `maximumBufferSize` as long as it is read before returning `.needMoreData`. I'm not sure if that was intended though. It does mean that if you get a large buffer and process many messages out of it immediately, you don't throw. This seems reasonable since the memory was already allocated anyway in the `ByteBuffer` given to the handler, although it may be copied in `startProcessing`.

A more correct copy of the `ByteToMessageHandler` behavior for `SingleStepByteToMessageProcessor` would be to place this check at the end of `decodeLoop`. I propose to do this unless you feel strongly this should be in `append`.

Reply:
I'm on the fence. @weissi?